### PR TITLE
WASI, Node API, Test Changes

### LINKED ISSUE #440: Move Nodeinfo (e.g. peer id, listener addrs, connections) to /node endpoint.
# Summary

For the workshop, we conveyed some of this info through `/health`, but this should be available from a separate endpoint (as a lot of it is really dynamic anyway). 

### static
- [ ] peer id

### dynamic
- [ ] listener addresses
- [ ] open connections

### LINKED ISSUE #498: Wasm: Setup for Wasm/Wasi component execution
# Summary

As we move toward effects (like #189) and having basic concepts like logging within our Guest Wasm modules, we need the ability to provide host interaction via Wasi. 

## Components 

- [ ] Can execute Wasm/Wasi compiled modules.
- [ ] Allow for the use of Wit interfaces provided by Wasi proposals. 
- [ ] Provide a homestar (host) package for others to use (including example/test-driven `homestar-functions`).

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
# Summary

A smattering of items mainly focused on WASI integration, test updates, and Node API changes. 

## Components
- [X] Closes #440 (with integration tests for dynamic data check)
- [X] Closes #495 
- [X] Closes #493 
- [X] Closes #494 
- [X] Closes #484 
- [X] Closes #498 
- [X] Fully-basked wit/wasi integration and incorporating [wit-deps](https://github.com/bytecodealliance/wit-deps) for shared packages
- [X] Allows for the execution of Wasi Wasm components, providing some basics around print, get-current-time, and logging (the latter using the WASI proposal's interface
- [X] CID change(s) for Wasm
- [X] Remove the need for manually killing/stopping procs in Homestar-runtime integration tests with `ChildGuard` wrapper util
- [X] New basic wasm mods (incorporating host functions and logging)
- [X] Parallelize the integration tests 🐕
- [X] Health check now validates running DB

### CODE REVIEW

**Comment by zeeshanlakhani** [comment]
nightly lint errors are being fixed in mainline clippy: https://github.com/rust-lang/rust-clippy/issues/12133.

File: homestar-runtime/tests/metrics.rs [hunk 1]
```diff
@@ -55,9 +51,9 @@ fn test_metrics_serial() -> Result<()> {
         .stdout(Stdio::piped())
         .spawn()
         .unwrap();
+    let _guard = ChildGuard::new(homestar_proc);
```
Comment (bgins) [inline homestar-runtime/tests/metrics.rs:54]: Not observing one, but is there any possible issue with `_guard` shadowing the file guard with the same name a few lines up?

File: homestar-runtime/tests/utils.rs [hunk 2]
```diff
@@ -31,7 +32,101 @@ pub(crate) const BIN_NAME: &str = "homestar";
...
+        }
+    }
+
+    /// Take the [Child] process from the [ChildGuard] and return the [ChildStdout].
+    pub(crate) fn stdout(&mut self) -> Option<ChildStdout> {
+        match &mut self.guard {
+            Some(child) => child.stdout.take(),
+            None => None,
+        }
+    }
```
Comment (bgins) [inline homestar-runtime/tests/utils.rs:65]: Nice! Love how these utilities came together. ❤️

File: homestar-runtime/tests/metrics.rs (hunk 1 above)
Comment (zeeshanlakhani) [reply homestar-runtime/tests/metrics.rs:54]: Nope.

File: homestar-runtime/src/settings/libp2p_config.rs [hunk 3]
```diff
@@ -128,6 +130,8 @@ impl Default for Libp2p {
         Self {
             announce_addresses: Vec::new(),
             dht: Dht::default(),
+            // https://github.com/libp2p/rust-libp2p/pull/4967
+            idle_connection_timeout: Duration::new(10, 0),
```
Comment (bgins) [inline homestar-runtime/src/settings/libp2p_config.rs:134]: Could we add this to `homestar-runtime/config/defaults.toml`?

File: homestar-runtime/src/network/swarm.rs [hunk 4]
```diff
@@ -127,7 +127,8 @@ pub(crate) async fn new(settings: &settings::Network) -> Result<Swarm<ComposedBe
             ),
         },
         peer_id,
-        swarm::Config::with_tokio_executor(),
+        swarm::Config::with_tokio_executor()
+            .with_idle_connection_timeout(settings.libp2p.idle_connection_timeout),
```
Comment (bgins) [inline homestar-runtime/src/network/swarm.rs:131]: Good catch, I think we lost this in an previous refactor. 👍

File: homestar-runtime/src/settings/libp2p_config.rs (hunk 3 above)
Comment (zeeshanlakhani) [reply homestar-runtime/src/settings/libp2p_config.rs:134]:
```toml
[node.network.libp2p]
idle_connection_timeout = 10
listen_address = "/ip4/0.0.0.0/tcp/0"
node_addresses = []
announce_addresses = []
transport_connection_timeout = 60
max_connected_peers = 32
max_announce_addresses = 10
```

File: homestar-wasm/src/wasmtime/world.rs [hunk 5]
```diff
@@ -13,45 +15,105 @@ use crate::{
...
+            wasi_ctx,
+            table,
+        }
     }
 
     /// Set fuel add.
-    pub fn add_fuel(&mut self, fuel: u64) {
+    pub fn set_fuel(&mut self, fuel: u64) {
         self.fuel = fuel
     }
```
Comment (bgins) [inline homestar-wasm/src/wasmtime/world.rs:112]: Minor: Update docs to remove "add"?

File: homestar-runtime/src/worker.rs [hunk 6]
```diff
@@ -562,66 +562,50 @@ mod test {
...
+                _ => panic!("Wrong event type"),
             }
-            _ => panic!("Wrong event type"),
-        };
+        }
 
-        assert!(rx.recv_async().await.is_err());
+        assert!(find_record);
+        assert!(get_providers);
+        assert!(captured_receipt);
```
Comment (bgins) [inline homestar-runtime/src/worker.rs:601]: Should there be a check on `receipts_cnt` here as well?

File: homestar-runtime/src/worker.rs [hunk 7]
```diff
@@ -562,66 +562,50 @@ mod test {
...
+                    let stored = workflow::Stored::default(Pointer::new(workflow_cid), 2);
+                    let mut info = workflow::Info::default(stored);
+                    info.increment_progress(receipt);
+                    let (_, workflow_info) =
+                        MemoryDb::get_workflow_info(workflow_cid, &mut conn).unwrap();
+                    assert_eq!(info.progress_count, workflow_info.progress_count);
+                    if receipts_cnt == 1 {
+                        captured_receipt = true;
+                    } else {
+                        receipts_cnt += 1;
```
Comment (bgins) [inline homestar-runtime/src/worker.rs:592]: Maybe don't need the conditional here? The receipts count can be incremented on each event and no harm is setting `captured_receipt` to true multiple times.

**Review by bgins — APPROVED** [review]
Looks great! The WASI Wasm components are a fantastic addition, and the tests are looking much nicer with the guards. 💯

**Comment by zeeshanlakhani** [comment]
failed tests are the nightlies which should be fixed up soon after the clippy false positive issue.

**Pull request merged**